x/ecocredit/server/core/send_test.go

```diff
@@ -0,0 +1,94 @@
+package core
+
+import (
+	"github.com/cosmos/cosmos-sdk/testutil/testdata"
+	"github.com/cosmos/cosmos-sdk/types"
+	"github.com/golang/mock/gomock"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+	"github.com/regen-network/regen-ledger/x/ecocredit/v1beta1"
+	"gotest.tools/v3/assert"
+	"testing"
+)
+
+func TestSend_Valid(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	_, _, recipient := testdata.KeyTestPubAddr()
+	s.setupClassProjectBatch(t)
+	any := gomock.Any()
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(_ interface{}, p *ecocredit.Params) {
+		p.AllowlistEnabled = false
+		p.CreditClassFee = types.NewCoins(types.NewInt64Coin("foo", 20))
+		p.CreditTypes = []*ecocredit.CreditType{{Name: "carbon", Abbreviation: "C", Unit: "tonne", Precision: 6}}
+	}).AnyTimes()
+
+	// s.Addr starting balance -> 10.5 tradable, 10.5 retired
+
+	_, err := s.k.Send(s.ctx, &v1beta1.MsgSend{
+		Sender:    s.addr.String(),
+		Recipient: recipient.String(),
+		Credits: []*v1beta1.MsgSend_SendCredits{
+			{BatchDenom: "C01-20200101-20210101-01", TradableAmount: "2.51"},
+			{BatchDenom: "C01-20200101-20210101-01", RetiredAmount: "1.30", RetirementLocation: "US-OR"},
+		},
+	})
+	assert.NilError(t, err)
+
+	senderBal, err := s.stateStore.BatchBalanceStore().Get(s.ctx, s.addr, 1)
+	assert.NilError(t, err)
+
+	recipientBal, err := s.stateStore.BatchBalanceStore().Get(s.ctx, recipient, 1)
+	assert.NilError(t, err)
+
+	// sender tradable -> 10.5 - 2.51 = 7.99
+	// recipient now has 2.51
+	// sender tradable -> 7.99 retires 1.30 = 6.69
+	// recipient now has 1.30 retired
+
+	assert.Equal(t, "6.69", senderBal.Tradable)
+	assert.Equal(t, "2.51", recipientBal.Tradable)
+	assert.Equal(t, "1.30", recipientBal.Retired)
+	assert.Equal(t, "10.5", senderBal.Retired) // retired credits should be untouched
+
+	sup, err := s.stateStore.BatchSupplyStore().Get(s.ctx, 1)
+	assert.NilError(t, err)
+
+	// retired -> 10.5 + 1.30(retired) = 11.80
+	// tradable -> 10.5 - 1.30 = 9.20
+
+	assert.Equal(t, "9.20", sup.TradableAmount)
+	assert.Equal(t, "11.80", sup.RetiredAmount)
+}
+
+func TestSend_Errors(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	_, _, recipient := testdata.KeyTestPubAddr()
+	s.setupClassProjectBatch(t)
+	any := gomock.Any()
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(_ interface{}, p *ecocredit.Params) {
+		p.AllowlistEnabled = false
+		p.CreditClassFee = types.NewCoins(types.NewInt64Coin("foo", 20))
+		p.CreditTypes = []*ecocredit.CreditType{{Name: "carbon", Abbreviation: "C", Unit: "tonne", Precision: 6}}
+	}).AnyTimes()
```

Comment:
Not sure if it's necessary to check times here.
```suggestion
	}).Times(0)
```